# Problem Set 1

In [114]:
import numpy as np

In [115]:
import pandas as pd

## 1. Lists

####  1.

In [116]:
movies = ["Interstellar", "Good Will Hunting", "The Accountant", 
          "Knives Out", "50/50", "Stand and Deliver"]

##### 2.

In [117]:
sliced_movies = movies[0:3]

#### 3.

In [118]:
first_last = movies[0:2] + movies[4:6]

## 2. Loops

#### 1.

In [119]:
numbers = []
for i in range(70,80):
    numbers.append(i)

#### 2.

In [120]:
sumInt = 0
for i in range(0,101):
    sumInt += i

#### 3.

In [121]:
names = pd.Series([33, 12, 45, 2, 17], index = ["Adam:", "Ashin:", "Inukai:", "Tanaka:", "Ikki:"])
names

Adam:      33
Ashin:     12
Inukai:    45
Tanaka:     2
Ikki:      17
dtype: int64

## 3. Comprehensions

#### 1.

In [122]:
def squares(first, last):
    return [i**2 for i in range(first, last + 1)]
print(squares(1, 10))

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


#### 2.

In [123]:
second = {i: i**2 for i in range(1,11)} 
second

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}

## 4. Functions and Data Transformations

#### 1.

In [124]:
time = 0
def numeric(time):
    minutes = time % 100
    hours = time // 100
    minutes = minutes/60
    final = str(int(hours) + float(minutes))
    return float(final)
print(numeric(1015))
print(numeric(345))

10.25
3.75


## 5. Load Data

#### 1.

In [125]:
gm = pd.read_csv("../Data/gapminder (1).csv.bz2", sep = "\t")

#### 2. a.

In [126]:
print(gm.shape)
print("There are 25 columns")

(13055, 25)
There are 25 columns


#### b.

In [127]:
print(gm.shape)
print("There are 13055 rows")

(13055, 25)
There are 13055 rows


#### c.

In [128]:
gm.head(3)

iso3   name iso2    region                       sub-region  \
0  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
1  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
2  ABW  Aruba   AW  Americas  Latin America and the Caribbean   

  intermediate-region    time  totalPopulation  fertilityRate  lifeExpectancy  \
0           Caribbean  1960.0          54211.0          4.820          65.662   
1           Caribbean  1961.0          55438.0          4.655          66.074   
2           Caribbean  1962.0          56225.0          4.471          66.444   

   ...  accessElectricity  agriculturalLand  agricultureTractors  \
0  ...                NaN               NaN                  NaN   
1  ...                NaN              20.0                  NaN   
2  ...                NaN              20.0                  NaN   

   cerealProduction  fertilizerHa        co2  greenhouseGases     co2_PC  \
0               NaN           NaN  11092.675              NaN  204.62037   
1               NaN           NaN  11576.719              NaN  208.82281   
2               NaN           NaN  12713.489              NaN  226.11808   

   pm2.5_35  battleDeaths  
0       NaN           NaN  
1       NaN           NaN  
2       NaN           NaN  

[3 rows x 25 columns]

Yes, this data does look reasonable.

## 6.

#### 1.

In [129]:
countries = gm.groupby("name").sum()["time"]   #This turns the df into a series which makes it easy to sum unique countries by using index.size
print(countries.index.size)

249


#### 2.

In [130]:
year = gm["time"]
print(year.max())
print(year.min())

2019.0
1960.0


Earliest year is 1960, latest or most recent year is 2019.

#### 3.

In [131]:
gm2 = gm[~gm.GDP_PC.isna()] 
gm2.time.max()

2019.0

2019 means that the data we have about GDP from 2019 is the most recent data on GDP.

#### 4.

In [132]:
gm3 = gm2.groupby("time").mean("GDP_PC")
wealth = gm3["GDP_PC"]
print(wealth.iloc[-1])

15055.352217243244


The number printed here shows the average wealth on this planet in the year 2019

#### 5.

In [133]:
gm4 = gm2.groupby("name").time.max()
print(gm4.nsmallest(n=5))

name
Djibouti                    2010.0
Faroe Islands               2010.0
Liechtenstein               2010.0
Turks and Caicos Islands    2010.0
Eritrea                     2011.0
Name: time, dtype: float64


The countries printed here are the 5 with the most number of recent years missing. By setting gm4 to the latest year recorded for each country,
I was then able to see what the smallest-latest year recorded was.

#### 6.

In [134]:
gm5 = gm2[gm2["time"] == 2019.0] #subsets data to only 2019 values
gm6 = gm5.groupby("region").mean()["GDP_PC"] #finds the mean GDP per capita in each continent
print((gm6).sort_values(ascending=False)) #shows the data in descending order

region
Europe      33227.983767
Americas    14516.826303
Asia        14116.124163
Oceania     11817.735554
Africa       2765.708956
Name: GDP_PC, dtype: float64


After subsetting to only the year 2019, I found the average GDP per capita across each continent by using the mean function, which is what
the numbers printed out tell us, in a decreasing order.

#### 7.

In [135]:
gm7 = gm2[gm2["time"] == 2019.0]
gm7 = gm7[["region", "name", "totalPopulation", "GDP_PC"]] #subsets to only region, name, and totalPopulation columns
gdp = gm7["GDP_PC"]
max_index = gdp.idxmax() #finds the index of the max life expectancy
min_index = gdp.idxmin() #finds the index of the min life expectancy
g1 = gm7.loc[max_index] 
g2 = gm7.loc[min_index]
print(g1)
print(g2)

region                  Europe
name                Luxembourg
totalPopulation       620001.0
GDP_PC             111043.5317
Name: 7039, dtype: object
region                 Africa
name                  Burundi
totalPopulation    11530580.0
GDP_PC              208.07473
Name: 843, dtype: object


By subsetting the data frame for only 2019 values, and finding the index of the smallest and largest GDP_PC values, I was then
able to traverse the data to find what country corresponds with said indicies, and print out data like population, GDP, and of course the 
names of the country/continent.

#### 8.

In [136]:
gm7 = gm2[gm2["time"] == 2019.0]
gm7 = gm7[["region", "name", "totalPopulation", "GDP_PC"]]  #subsets to only region, name, and totalPopulation columns
gm9 = gm7.groupby("region").max()
gdp = gm7["GDP_PC"]
for i in range(len(gm7["region"])): 
    max_index = gdp.idxmax() #finds the index of the max life expectancy
    min_index = gdp.idxmin() #finds the index of the minimum life expectancy
    g1 = gm7.iloc[i]
    g2 = gm7.iloc[i]
pd.concat((gm9, g1, g2), axis = 1) #combines data

name  \
Africa                                                    Zimbabwe   
Americas                                                   Uruguay   
Asia                                                         Yemen   
Europe           United Kingdom of Great Britain and Northern I...   
Oceania                                                    Vanuatu   
region                                                         NaN   
name                                                           NaN   
totalPopulation                                                NaN   
GDP_PC                                                         NaN   

                 totalPopulation        GDP_PC       13054       13054  
Africa              2.009636e+08   15048.74693         NaN         NaN  
Americas            3.282395e+08   94106.36125         NaN         NaN  
Asia                1.397715e+09   63281.67468         NaN         NaN  
Europe              1.444063e+08  111043.53170         NaN         NaN  
Oceania             2.536574e+07   57183.37935         NaN         NaN  
region                       NaN           NaN      Africa      Africa  
name                         NaN           NaN    Zimbabwe    Zimbabwe  
totalPopulation              NaN           NaN  14645468.0  14645468.0  
GDP_PC                       NaN           NaN  1183.09882  1183.09882

This table should show the richest and poorest countries in each continent. I was not able to get it to show correctly though, my problem was with trying to show the data for each continent, neither my groupby nor my for loop attempts worked :(.

#### 9.

Although I don't have the table working correctly, I did some research and got outside data from: https://worldpopulationreview.com/country-rankings/richest-countries-in-north-america

Most of the placings for each country makes sense to me, however like mentioned, Bermuda being the richest country in the Americas surprised me. This is because a lot of people (like myself) don't even consider Bermuda when thinking about GDP per capita. Bermuda's high GDP number pobably come from all the tourism and of course, how small it is. I think that it makes sense for most of rich countries to have a small population when looking at GDP per capita because if you sum up all of the money and divide it by a smaller population rather than a large one, you get a bigger number in return.

## 7.

#### 1.

In [137]:
hm = gm[gm["time"] == 1960.0] #sets data to look only at 1960
hm = hm[["lifeExpectancy", "name"]]  #subsets to only name, and lifeExcpectancy columns
hm.groupby("name").mean()["lifeExpectancy"]
print(hm.isna().sum()) #finds the number of NA's in life expectancy in the data
hm2 = gm[gm["time"] == 2019.0] #sets data to look only at 2019
hm2 = hm2[["lifeExpectancy", "name"]]
hm2.groupby("name").mean()["lifeExpectancy"]
print(hm2.isna().sum()) #finds the number of NA's in life expectancy in the data

lifeExpectancy    27
name               4
dtype: int64
lifeExpectancy    19
name               4
dtype: int64


It appears that in 1960 there were 27 cases where life expectancy wasn't reported, and in 2019 there were 19 cases where life expectancy wasn't reported.

#### 2.

In [138]:
hm3 = gm[["time", "name", "lifeExpectancy"]]  #subsets to only time, name, and lifeExcpectancy columns
le = hm3["lifeExpectancy"]
max_index = le.idxmax() #finds the index of the max life expectancy
min_index = le.idxmin() #finds the index of the min life expectancy
print(hm3.loc[max_index])
print(hm3.loc[min_index])

time                  2012.0
name              San Marino
lifeExpectancy      85.41707
Name: 10581, dtype: object
time                1977.0
name              Cambodia
lifeExpectancy      18.907
Name: 6097, dtype: object


The values printed here show the country, year, and number for the highest/lowest life expectancies. 

#### 3.

The number 18.907 corresponds to the Cambodian genocide, which happened throughout the 1970's.



Reference: https://cla.umn.edu/chgs/holocaust-genocide-education/resource-guides/cambodia#:~:text=Lasting%20for%20four%20years%20(between,following%20the%20Cambodian%20Civil%20War.

#### 4.

In [139]:
hd7 = gm2[gm2["time"] == 2019.0]
hd7 = hd7[["region", "name", "lifeExpectancy"]] #subsets to only region, name, and lifeExcpectancy columns
hd8 = hd7.groupby("region").max() #shows the max life expectancy of each region
print(hd8)
hd9 = hd7.groupby("region").min() #shows the min life expectancy of each region
print(hd9)

                                                       name  lifeExpectancy
region                                                                     
Africa                                             Zimbabwe        76.88000
Americas                                            Uruguay        82.04878
Asia                                                  Yemen        85.07805
Europe    United Kingdom of Great Britain and Northern I...        83.70488
Oceania                                             Vanuatu        82.90000
                         name  lifeExpectancy
region                                       
Africa                Algeria        53.28300
Americas  Antigua and Barbuda        64.00100
Asia              Afghanistan        64.83300
Europe                Albania        71.82732
Oceania             Australia        64.50100


This code prints out the countries with the shortest and longest life expectancies for each continent.

#### 5.

In [140]:
hm7 = gm[["time", "name", "lifeExpectancy"]] #subsets to only time, name, and lifeExcpectancy columns
le = hm3["time"]
min_index = le.idxmin() #finds the index of the minimum life expectancy
print(hm3.loc[min_index])

time              1960.0
name               Aruba
lifeExpectancy    65.662
Name: 0, dtype: object


This table is meant to show the first valid Le and its corresponding year. Once again, I was unable to print the values for more than 1 country(aruba). Tried to use groupby and a for loop to print out the values for each country instead of just one, however I was unsuccessful in my attempt. 

#### 6.

In [141]:
hm9 = gm[["time", "name", "lifeExpectancy"]] #subsets to only time, name, and lifeExcpectancy columns
le = hm3["lifeExpectancy"]
max_index = le.idxmax() #finds the index of the max life expectancy
print(hm3.loc[max_index])

time                  2012.0
name              San Marino
lifeExpectancy      85.41707
Name: 10581, dtype: object


This table is meant to show the most recent valid Le and its corresponding year. Once again, I was unable to print the values for more than 1 country(aruba). Tried to use groupby and a for loop to print out the values for each country instead of just one, however I was unsuccessful in my attempt. 

#### 7.

In [142]:
bruh1 = hd9[["name", "time", "lifeExpectancy"]].set_index("name") #subsets data frame for shortest life expectancies 
bruh2 = hd8[["name", "time", "lifeExpectancy"]].set_index("name") # ' ' but opposite
LE1 = bruh1.lifeExpectancy #first valid LE
LE2 = bruh2.lifeExpectancy # ' '
time1 = bruh1.time #year of first valid LE
time2 = bruh2.time # ' '
n = bruh2 - bruh1 #this and below are calculations for the GR formula
gr = 100 * ((LE2 / LE1)**(1 / n) - 1)
gr1 = pd.concat([time2, LE2, time1, LE1, gr], axis = 1).rename({0: "growthRate"}, axis = 1)
fast = gr1.sort_values("growthRate", ascending = False).head(10) #combining data into dataframe 
slow = gr1.sort_values("growthRate", ascending = False).head(10)
print(fast)
print(slow)

KeyError: "['time'] not in index"

This table was meant to show the 10 fastest and 10 slowest growing countries in the world based off the growth rate equation. However, since I was unable to solve the previous problems, I had trouble solving this one, and was unable to get it. I got help for making the data table off of teams, but I was still not able to completely figure it out. 

#### 8.

Although I don't have the data frame output, after using data from https://www.statista.com/statistics/264687/countries-with-the-highest-population-growth-rate/ I was able to find a common theme among the fastest growing countries. The growth rate may not be the same, but I think it would share a common theme of the fastest growing countries being relatively poor. Although studies show that increased life expectancies lead to higher population growth, so I am a little confused here, becuase a lot of the poor countries shown don't have very high life expectancies. Undoubtedly though, life expectancies have increased all around the world since their first record. 

I spent about 10 hours working on this assignment, should've done it more efficiently and gone to office hours, but I was busy and had to end up using a late day.